src/modules/adaptors/Frax/DebtFTokenAdaptorV2.sol

```diff
@@ -0,0 +1,300 @@
+     */
+    function assetOf(bytes memory adaptorData) public view override returns (ERC20) {
+        IFToken fraxlendPair = abi.decode(adaptorData, (IFToken));
+        return ERC20(fraxlendPair.asset());
+    }
+
+    /**
+     * @notice This adaptor reports values in terms of debt.
+     */
+    function isDebt() public pure override returns (bool) {
+        return true;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    // `borrowAsset`
+    /**
+     * @notice Allows strategists to borrow assets from Fraxlend.
+     * @notice `debtTokenToBorrow` must be the debtToken, NOT the underlying ERC20.
+     * @param fraxlendPair the Fraxlend Pair to borrow from.
+     * @param amountToBorrow the amount of `debtTokenToBorrow` to borrow on Fraxlend.
+     * NOTE: `borrowAsset` is the same btw v1 and v2 FraxlendPairs
+     */
+    function borrowFromFraxlend(IFToken fraxlendPair, uint256 amountToBorrow) public {
+        bytes32 positionHash = keccak256(abi.encode(identifier(), true, abi.encode(address(fraxlendPair))));
+        uint32 positionId = Cellar(address(this)).registry().getPositionHashToPositionId(positionHash);
+        if (!Cellar(address(this)).isPositionUsed(positionId))
+            revert DebtFTokenAdaptor__FraxlendPairPositionsMustBeTracked(address(fraxlendPair));
+
+        _borrowAsset(amountToBorrow, fraxlendPair);
+
+        // Check health factor is still satisfactory
+        uint256 _exchangeRate = _updateExchangeRate(fraxlendPair);
+        // Check if borrower is insolvent after this borrow tx, revert if they are
+        if (minimumHealthFactor > (_isSolvent(fraxlendPair, _exchangeRate))) {
+            revert DebtFTokenAdaptor__HealthFactorTooLow(address(fraxlendPair));
+        }
+    }
+
+    // `repayDebt`
+
+    /**
+     * @notice Allows strategists to repay loan debt on Fraxlend Pair. Make sure to call addInterest() beforehand to ensure we are repaying what is required.
+     * @dev Uses `_maxAvailable` helper function, see BaseAdaptor.sol
+     * @param _fraxlendPair the Fraxlend Pair to repay debt from.
+     * @param _debtTokenRepayAmount the amount of `debtToken` to repay with.
+     */
+    function repayFraxlendDebt(IFToken _fraxlendPair, uint256 _debtTokenRepayAmount) public {
+        ERC20 tokenToRepay = ERC20(_fraxlendPairAsset(_fraxlendPair));
+        uint256 debtTokenToRepay = _maxAvailable(tokenToRepay, _debtTokenRepayAmount);
+        uint256 sharesToRepay = _toAssetShares(_fraxlendPair, debtTokenToRepay, false, true);
+        uint256 sharesAccToFraxlend = _userBorrowShares(_fraxlendPair); // get fraxlendPair's record of borrowShares atm
+        if (sharesAccToFraxlend == 0) revert DebtFTokenAdaptor__CannotRepayNoDebt(address(_fraxlendPair)); // NOTE: from checking it out, unless `userBorrowShares[_borrower] -= _shares;` reverts, then fraxlendCore lets users repay FRAX w/ no limiters.
```

Comment:
userBorrowShares should revert from underflow if you tried repaying more than owed

Reply:
Ah okay, so remove this check?

Reply:
We discussed this, it's understood but we are leaving it so it is explicit for the strategist.